packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx

```diff
@@ -0,0 +1,228 @@
+import * as AI from '../../../index.js';
+import { ConversationMessage, FunctionResponse, FunctionResponseProps } from '../../../core/conversation.js';
+
+import { RecursiveCharacterTextSplitter } from 'langchain/text_splitter';
+import { getEncoding } from 'js-tiktoken';
+import yaml from 'js-yaml';
+import _ from 'lodash';
+import { UseToolsProps } from '../../use-tools.js';
+import { cohereContext, MarkdownChunkFormatter, RerankerFormatted } from '../../../lib/cohere.js';
+
+export type LengthFunction = ((text: string) => number) | ((text: string) => Promise<number>);
+export interface RedactedFuncionResponseMetadata {
+  isRedacted: true;
+  chunks: string[];
+}
+
+const getEncoder = _.once(() => getEncoding('cl100k_base'));
+
+function openAITokenCount(text: string) {
+  return getEncoder().encode(text).length;
+}
+
+export async function TruncateByChars(
+  {
+    children,
+    maxLength,
+  }: {
+    children: AI.Node;
+    maxLength: number;
+  },
+  { render }: AI.ComponentContext
+) {
+  const stringified = await render(children);
+  if (stringified.length <= maxLength) {
+    return stringified;
+  }
+  return `${stringified.slice(0, maxLength - 3)}...`;
+}
+
+export interface LargeFunctionResponseProps {
+  maxLength: number;
+  failedMaxLength: number;
+  numChunks: number;
+  lengthFunction?: LengthFunction;
+}
+
+async function LargeFunctionResponseHandler(
+  {
+    children,
+    maxLength = 4000,
+    failedMaxLength = 1000,
+    numChunks = 4,
+    // use Cohere token counter?
+    lengthFunction = openAITokenCount,
+    ...props
+  }: AI.PropsOfComponent<typeof FunctionResponse> & LargeFunctionResponseProps,
+  { render, logger, getContext }: AI.ComponentContext
+) {
+  if (props.failed) {
+    return (
+      // TODO: fix issue between maxLength chars and tokens
+      <FunctionResponse {...props}>
+        <TruncateByChars maxLength={failedMaxLength}>{children}</TruncateByChars>
+      </FunctionResponse>
+    );
+  }
+
+  let stringified = await render(children);
+
+  // Option 1: do nothing if it's already small enough
+  if ((await lengthFunction(stringified)) <= maxLength) {
+    return <FunctionResponse {...props}>{stringified}</FunctionResponse>;
+  }
+
+  stringified = yamlOptimizeIfPossible(stringified);
+
+  // Option 2: try dumping as YAML, if it's small enough then we are done
+  if ((await lengthFunction(stringified)) <= maxLength) {
+    return <FunctionResponse {...props}>{stringified}</FunctionResponse>;
+  }
+
+  // Option 3 (last reosrt): split into chunks and allow LLM to query by similarity
+  // Requires Cohere API key for doing similarity search
```

Comment:
Hmm. I feel like, if the Cohere API key isn't set, it would be better to just truncate the response.

Reply:
> I feel like, if the Cohere API key isn't set

We do set it using a context inside `request-wrapper.tsx`, so Fixie sidekicks will have it set. Also we do have a default context that uses `COHERE_API_KEY` env var. Is there anywhere I should document this?

Reply:
cc @hessamb

Reply:
I think that's a good idea. But falling back to truncate by char is also a good interim solution for hackathon (since we expect cohere API to be set anyway).

Reply:
Done.